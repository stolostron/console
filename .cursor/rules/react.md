# React Rules for ACM Console

## Component Patterns

### Functional Components
- Use functional components with hooks exclusively
- Implement proper prop interfaces with TypeScript
- Use React.memo for expensive components
- Follow established naming conventions (PascalCase)

### Hook Usage
- Use custom hooks for business logic separation
- Implement proper dependency arrays in useEffect
- Use useCallback for event handlers passed as props
- Leverage useMemo for expensive computations

### State Management
- Use Recoil atoms through useSharedAtoms() pattern
- Access atoms with useRecoilValue() for reading
- Use useRecoilValueGetter() for async operations
- Follow established atom naming conventions

Example (based on actual codebase patterns):
```typescript
/* Copyright Contributors to the Open Cluster Management project */
import { useMemo } from 'react'
import { useRecoilValue } from 'recoil'
import { useTranslation } from '../lib/acm-i18next'
import { useSharedAtoms } from '../shared-recoil'
import { AcmTable } from '../ui-components'

export function ClusterList() {
  const { t } = useTranslation()
  const { managedClustersState } = useSharedAtoms()
  const clusters = useRecoilValue(managedClustersState)
  
  const filteredClusters = useMemo(
    () => clusters.filter(cluster => cluster.status?.conditions),
    [clusters]
  )
  
  return (
    <AcmTable
      items={filteredClusters}
      columns={clusterColumns}
      keyFn={cluster => cluster.metadata.uid}
    />
  )
}
```

### Conditional Rendering
- Use logical AND (`&&`) for simple conditional rendering (ACM standard pattern):
  ```typescript
  {condition && <Component />}
  {props.additionalLabels && (<span><AcmLabels /></span>)}
  {expandable && itemCount > columnCount && (<AcmButton />)}
  ```
- Use ternary operators for either/or scenarios:
  ```typescript
  {props.isCompact ? <Popover /> : renderLabelGroup()}
  {showAll ? t('Show less') : t('Show all')}
  ```
- Early returns for major conditional logic:
  ```typescript
  if (props.hidden) return <></>
  if (!data) return <LoadingState />
  ```

### Performance
- Use `React.memo` for components that receive the same props frequently
- Use `useMemo` for expensive calculations
- Use `useCallback` for functions passed to child components
- Implement lazy loading for heavy components
- Avoid inline object/array creation in render
- Avoid using index as the key prop value
- Avoid using inline functions in JSX or TSX blocks
- Use proper key props for list items

### Error Boundaries
- Implement error boundaries for route-level components
- Use proper error state management
- Provide user-friendly error messages
- Log errors for debugging purposes

### Accessibility
- Use semantic HTML elements
- Implement proper ARIA attributes
- Ensure keyboard navigation works
- Test with screen readers

### Component Organization
- Keep components small and focused on a single responsibility
- Use PascalCase for component names and component files
- Use descriptive component names that indicate the component's purpose
- Avoid abbreviations in component names
- Separate presentation from business logic
- Use proper file structure (components/, ui-components/, routes/)
- Export components properly with named exports
- Avoid props drilling where props are only passed down to child components
- Do not mutate props or state - create a copy array or object if possible
