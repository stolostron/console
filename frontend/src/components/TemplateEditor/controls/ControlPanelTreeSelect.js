'use strict'

/* eslint-disable react/no-unused-state, react/no-unused-prop-types, jsx-a11y/autocomplete-valid */

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import TimesCircleIcon from '@patternfly/react-icons/dist/js/icons/times-circle-icon'
import ControlPanelFormGroup from './ControlPanelFormGroup'
import cloneDeep from 'lodash/cloneDeep'
import get from 'lodash/get'

class ControlPanelTreeSelect extends React.Component {
  static propTypes = {
    control: PropTypes.object,
    controlId: PropTypes.string,
    handleChange: PropTypes.func,
  }

  static getDerivedStateFromProps(props, state) {
    const { control, handleChange } = props
    const handleTreeChange = (evt) => {
      control.active = evt.selectedItem
      handleChange(evt)
    }

    const { available = [] } = control
    const { branches = 0 } = state
    let { active } = control
    let { currentSelection, searchList, indexes = [], isOpen, searchText } = state
    const branchLabels = []
    let currentAvailable = []

    // clicked on a branch in search mode, search for that branch
    if (currentSelection !== undefined && searchText) {
      const { value, branch } = searchList[currentSelection]
      if (!value) {
        searchText = branch
        currentSelection = undefined
      }
    }

    /////////////////////////////////////////////////////////////
    // search mode
    if (searchText && searchText.length) {
      // nothing selected, filter list
      if (currentSelection === undefined) {
        searchList = []
        const findText = searchText.toLowerCase()
        const fillArray = (arry, branchMatch, indent) => {
          arry.forEach(({ label, children, value, description }) => {
            if (value) {
              const instance = `${value} - ${description}`
              if (branchMatch || instance.toLowerCase().indexOf(findText) !== -1) {
                currentAvailable.push({ instance, indent })
                searchList.push({ value, description })
              }
            } else if (children) {
              const beg = currentAvailable.length
              const bm = branchMatch || label.toLowerCase().indexOf(findText) !== -1
              fillArray(children, bm, indent + 20)
              if (currentAvailable.length > beg) {
                currentAvailable.splice(beg, 0, { branch: label, indent })
                searchList.splice(beg, 0, { branch: label })
              }
            }
          })
        }
        fillArray(available, false, 0)
        isOpen = true
      } else {
        // handle change
        const { value, description } = searchList[currentSelection]
        active = `${value}  # ${description}`
        handleTreeChange({ selectedItem: active })
        currentAvailable = []
        indexes = []
        isOpen = false
        searchText = null
      }
    } else {
      /////////////////////////////////////////////////////////////
      // tree mode
      // get current list using indexes
      if (currentSelection !== undefined) {
        currentSelection -= branches
        if (currentSelection >= 0) {
          indexes = cloneDeep(indexes)
          indexes.push(currentSelection)
        } else {
          // clicked a branch label
          indexes = indexes.slice(0, currentSelection)
        }
      }

      let path = indexes.map((index) => `[${index}]`).join('.children')
      currentAvailable = path ? get(available, path) : available
      currentAvailable = currentAvailable.children || currentAvailable
      let indent = 0
      if (Array.isArray(currentAvailable)) {
        path = ''
        indexes.forEach((index) => {
          path += `[${index}]`
          let label = get(available, `${path}.label`)
          if (label) {
            label = `${label}`
            branchLabels.push({ branch: label, indent })
            path += '.children'
            indent += 20
          }
        })
        currentAvailable = [
          ...branchLabels,
          ...currentAvailable.map((item) => {
            if (item.label) {
              return { branch: item.label, indent }
            } else {
              return {
                instance: `${item.value} - ${item.description}`,
                indent,
              }
            }
          }),
        ]
      } else {
        // handle change
        active = `${currentAvailable.value}  # ${currentAvailable.description}`
        handleTreeChange({ selectedItem: active })
        currentAvailable = []
        indexes = []
        searchText = null
        isOpen = false
      }
    }
    return {
      active,
      currentAvailable,
      currentSelection: undefined,
      indexes,
      branches: branchLabels.length,
      isOpen,
      searchText,
      searchList,
    }
  }

  constructor(props) {
    super(props)
    this.state = {
      isOpen: false,
      searchText: null,
    }
    // create active map
    this.addAvailableMap(props)
    this.onDocClick = (event) => {
      const clickedOnToggle = this.parentRef && this.parentRef.contains(event.target)
      const clickedWithinMenu = this.menuRef && this.menuRef.contains && this.menuRef.contains(event.target)
      const clickedWithinClear = this.clearRef && this.clearRef.contains && this.clearRef.contains(event.target)
      const clickedWithinToggle = this.toggleRef && this.toggleRef.contains && this.toggleRef.contains(event.target)
      if (this.state.isOpen && !(clickedOnToggle || clickedWithinMenu || clickedWithinClear || clickedWithinToggle)) {
        this.setState({ isOpen: false })
      }
    }
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.onDocClick)
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.onDocClick)
  }

  setParentRef = (ref) => {
    this.parentRef = ref
  }

  setMenuRef = (ref) => {
    this.menuRef = ref
  }

  setClearRef = (ref) => {
    this.clearRef = ref
  }

  setToggleRef = (ref) => {
    this.toggleRef = ref
  }

  addAvailableMap(props) {
    const { control } = props
    control.availableMap = {}
    const { available, availableMap } = control
    this.addAvailableMapHelper(available, availableMap)
  }

  addAvailableMapHelper(available = [], availableMap) {
    available.forEach(({ children, value, description }) => {
      if (children) {
        this.addAvailableMapHelper(children, availableMap)
      } else {
        availableMap[value] = description
      }
    })
  }

  render() {
    const { controlId, control } = this.props
    const { name, availableMap = {}, exception, disabled } = control
    const { isOpen, active, currentAvailable, indexes, searchText } = this.state
    const currentActive = availableMap[active] ? `${active} - ${availableMap[active]}` : active

    const toggleClasses = classNames({
      'tf--list-box__menu-icon': true,
      'tf--list-box__menu-icon--open': isOpen,
    })

    const aria = isOpen ? 'Close menu' : 'Open menu'
    const key = `${controlId}-${name}-${currentAvailable
      .map(({ branch, instance }) => {
        return branch || instance
      })
      .join('-')}`
    const validated = exception ? 'error' : undefined
    const inputClasses = classNames({
      'pf-c-form-control': true,
      input: true,
      disabled: disabled,
    })
    return (
      <React.Fragment>
        <div className="creation-view-controls-treeselect">
          <ControlPanelFormGroup controlId={controlId} control={control}>
            <div id={controlId}>
              <div role="listbox" aria-label="Choose an item" tabIndex="0" className="tf--list-box">
                <div
                  role="button"
                  className={inputClasses}
                  tabIndex="0"
                  type="button"
                  aria-label={aria}
                  aria-expanded={isOpen}
                  aria-haspopup="true"
                  data-toggle="true"
                  onClick={this.clickToggle.bind(this)}
                  onKeyPress={this.pressToggle.bind(this)}
                >
                  <input
                    className="pf-c-combo-control"
                    aria-label="ListBox input field"
                    ref={this.setParentRef}
                    spellCheck="false"
                    role="combobox"
                    aria-controls={key}
                    disabled={disabled}
                    aria-expanded="true"
                    autoComplete="new-password"
                    id={`${controlId}-input`}
                    placeholder=""
                    style={validated === 'error' ? { borderBottomColor: 'red' } : undefined}
                    value={searchText !== null ? searchText : currentActive}
                    onFocus={(e) => {
                      e.target.select()
                    }}
                    onKeyDown={this.pressPress.bind(this)}
                    onChange={(evt) => this.setState({ searchText: evt.currentTarget.value })}
                    data-testid={`tree-${controlId}`}
                  />
                  {!disabled && (
                    <div
                      role="button"
                      className="tf--list-box__selection"
                      tabIndex="0"
                      style={{ color: '#6a6e73' }}
                      title="Clear selected item"
                      ref={this.setClearRef}
                      onClick={this.clickClear.bind(this)}
                      onKeyPress={this.pressClear.bind(this)}
                    >
                      <TimesCircleIcon aria-hidden />
                    </div>
                  )}
                  {!disabled && (
                    <div
                      role="button"
                      tabIndex="0"
                      className={toggleClasses}
                      ref={this.setToggleRef}
                      onClick={this.clickToggle.bind(this)}
                      onKeyPress={this.pressToggle.bind(this)}
                    >
                      <svg
                        fillRule="evenodd"
                        height="5"
                        role="img"
                        viewBox="0 0 10 5"
                        width="10"
                        alt={aria}
                        aria-label={aria}
                      >
                        <title>Close menu</title>
                        <path d="M0 0l5 4.998L10 0z" />
                      </svg>
                    </div>
                  )}
                </div>
                {!disabled && isOpen && (
                  <div className="tf--list-box__menu" key={key} id={key} ref={this.setMenuRef}>
                    {currentAvailable.map(({ branch, instance, indent = 0 }, inx) => {
                      const itemClasses = classNames({
                        'tf--list-box__menu-item': true,
                        'tf--list-box__menu-branch': branch,
                        searching: searchText,
                        open: inx < indexes.length,
                      })
                      const label = branch || instance
                      return (
                        <div
                          role="button"
                          key={label}
                          className={itemClasses}
                          id={`${controlId}-item-${inx}`}
                          tabIndex="0"
                          style={{
                            textIndent: `${indent}px`,
                            whiteSpace: 'pre',
                          }}
                          onClick={this.clickSelect.bind(this, inx)}
                          onKeyPress={this.pressSelect.bind(this, inx)}
                        >
                          {this.renderLabel(label, searchText)}
                        </div>
                      )
                    })}
                  </div>
                )}
              </div>
            </div>
          </ControlPanelFormGroup>
        </div>
      </React.Fragment>
    )
  }

  renderLabel(label, searchText) {
    const inx = searchText && searchText.length && label.toLowerCase().indexOf(searchText.toLowerCase())
    if (inx !== null && inx >= 0) {
      label = [label.substr(0, inx), label.substr(inx, searchText.length), label.substr(inx + searchText.length)]
      return (
        <React.Fragment>
          {label[0]}
          <b>{label[1]}</b>
          {label[2]}
        </React.Fragment>
      )
    } else {
      return <React.Fragment>{label}</React.Fragment>
    }
  }

  pressPress(e) {
    if (e.key === 'Escape') {
      this.clickClear()
    }
  }

  pressToggle(e) {
    if (e.key === 'Enter') {
      this.clickToggle()
    } else if (e.key === 'Escape') {
      this.clickClear()
    }
  }

  clickToggle(e) {
    if (e) {
      e.stopPropagation()
    }
    const clickedWithinClear = e && this.clearRef && this.clearRef.contains && this.clearRef.contains(e.target)
    const clickedWithinToggle = e && this.toggleRef && this.toggleRef.contains && this.toggleRef.contains(e.target)
    if (!(this.state.searchText || clickedWithinClear) || clickedWithinToggle) {
      this.setState((preState) => {
        let { currentAvailable, currentSelection, searchText, indexes, isOpen } = preState
        isOpen = !isOpen
        if (!isOpen) {
          currentAvailable = []
          currentSelection = undefined
          searchText = null
          indexes = []
        }
        return {
          currentAvailable,
          currentSelection,
          searchText,
          indexes,
          isOpen,
        }
      })
    }
  }

  pressSelect(inx, e) {
    if (e.key === 'Enter') {
      this.clickSelect(inx)
    }
  }

  clickSelect(inx) {
    this.setState({ currentSelection: inx })
  }

  pressClear(inx, e) {
    if (e && e.key === 'Enter') {
      this.clickClear()
    }
  }

  clickClear() {
    this.setState({ searchText: '' })
    const { control, handleChange } = this.props
    control.active = ''
    handleChange()
  }
}

export default ControlPanelTreeSelect
