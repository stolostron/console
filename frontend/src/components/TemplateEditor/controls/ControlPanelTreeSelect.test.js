/* Copyright Contributors to the Open Cluster Management project */
'use strict'

import React from 'react'
import ControlPanelTreeSelect from './ControlPanelTreeSelect'
import { render } from '@testing-library/react'

export const control = {
    active: false,
    name: 'creation.app.name',
    tooltip: 'tooltip.creation.app.name',
    controlData: [],
    id: 'checkbox',
    type: 'checkbox',
}

describe('ControlPanelTreeSelect component', () => {
    it('renders as expected', () => {
        const fn = jest.fn()

        const Component = () => {
            return (
                <ControlPanelTreeSelect
                    key={'key'}
                    control={control}
                    controlId={'controlId'}
                    handleChange={fn}
                    i18n={fn}
                />
            )
        }
        const { asFragment } = render(<Component />)
        expect(asFragment()).toMatchSnapshot()
    })
})
