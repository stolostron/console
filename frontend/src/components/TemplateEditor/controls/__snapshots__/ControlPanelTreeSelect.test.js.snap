// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlPanelTreeSelect component renders as expected 1`] = `
<DocumentFragment>
  <div
    class="creation-view-controls-treeselect"
  >
    <div
      class="pf-c-form__group"
      id="controlId-label"
    >
      <div
        class="pf-c-form__group-label"
      >
        <label
          class="pf-c-form__label"
          for="controlId"
        >
          <span
            class="pf-c-form__label-text"
          >
            creation.app.name
          </span>
        </label>
         
        <button
          aria-label="More info"
          class="pf-c-form__group-label-help"
          id="controlId-label-help-button"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
            />
          </svg>
        </button>
      </div>
      <div
        class="pf-c-form__group-control"
      >
        <div>
          <div
            id="controlId"
          >
            <div
              aria-label="Choose an item"
              class="tf--list-box"
              role="listbox"
              tabindex="0"
            >
              <div
                aria-expanded="false"
                aria-haspopup="true"
                aria-label="Open menu"
                class="pf-c-form-control input"
                data-toggle="true"
                role="button"
                tabindex="0"
                type="button"
              >
                <input
                  aria-controls="controlId-creation.app.name-"
                  aria-expanded="true"
                  aria-label="ListBox input field"
                  autocomplete="new-password"
                  class="pf-c-combo-control"
                  data-testid="tree-controlId"
                  id="controlId-input"
                  placeholder=""
                  role="combobox"
                  spellcheck="false"
                  value="false"
                />
                <div
                  class="tf--list-box__selection"
                  role="button"
                  style="color: rgb(106, 110, 115);"
                  tabindex="0"
                  title="Clear selected item"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style="vertical-align: -0.125em;"
                    viewBox="0 0 512 512"
                    width="1em"
                  >
                    <path
                      d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
                    />
                  </svg>
                </div>
                <div
                  class="tf--list-box__menu-icon"
                  role="button"
                  tabindex="0"
                >
                  <svg
                    alt="Open menu"
                    aria-label="Open menu"
                    fill-rule="evenodd"
                    height="5"
                    role="img"
                    viewBox="0 0 10 5"
                    width="10"
                  >
                    <title>
                      Close menu
                    </title>
                    <path
                      d="M0 0l5 4.998L10 0z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
